vendor/github.com/newrelic/infra-integrations-sdk/jmx/jmx.go

```diff
@@ -191,62 +237,85 @@ func openConnection(config *connectionConfig) error {
 	return nil
 }
 
-// Close will finish the underlying nrjmx application by closing its standard
-// input and canceling the execution afterwards to clean-up.
-func Close() {
-	lock.Lock()
+func handleStdErr(ctx context.Context) {
+	scanner := bufio.NewReaderSize(cmdError, jmxLineInitialBuffer)
 
-	if cmd == nil {
-		lock.Unlock()
-		return
+	var line string
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			break
+		}
+
+		line, err = scanner.ReadString('\n')
+		// API needs re to allow stderr full read before closing
+		if err != nil && err != io.EOF && !strings.Contains(err.Error(), "file already closed") {
+			log.Error(fmt.Sprintf("error reading stderr from JMX tool: %s", err.Error()))
+		}
+		if strings.HasPrefix(line, "WARNING") {
+			msg := line[7:]
+			if strings.Contains(msg, "Can't parse bean name") {
+				cmdErrC <- ErrBeanPattern
+				return
+			}
+			cmdWarnC <- msg
+		}
+		if err != nil {
+			cmdErrC <- err
+			return
+		}
 	}
+}
 
+// Close will finish the underlying nrjmx application by closing its standard
+// input and canceling the execution afterwards to clean-up.
+func Close() {
 	cancel()
-	_ = cmdIn.Close()
-	_ = cmdError.Close()
-
-	lock.Unlock()
 
 	done.Wait()
 }
 
-func doQuery(ctx context.Context, out chan []byte, errorChan chan error, queryString []byte) {
-	lock.Lock()
+func doQuery(ctx context.Context, out chan []byte, queryErrC chan error, queryString []byte) {
 	if _, err := cmdIn.Write(queryString); err != nil {
-		lock.Unlock()
-		errorChan <- fmt.Errorf("writing query string: %s", err.Error())
+		queryErrC <- fmt.Errorf("writing nrjmx stdin: %s", err.Error())
 		return
 	}
 
-	scanner := bufio.NewScanner(cmdOut)
-	scanner.Buffer([]byte{}, jmxLineBuffer) // Override default buffer to increase buffer size
-	lock.Unlock()
+	scanner := bufio.NewReaderSize(cmdOut, jmxLineInitialBuffer)
 
-	if scanner.Scan() {
+	var b []byte
+	var err error
+	for {
 		select {
 		case <-ctx.Done():
 			return
-		case out <- scanner.Bytes():
 		default:
+			break
 		}
-	} else {
-		if err := scanner.Err(); err != nil {
-			errorChan <- fmt.Errorf("error reading output from JMX tool: %v", err)
-		} else {
-			// If scanner.Scan() returns false but err is also nil, it hit EOF. We consider that a problem, so we should return an error.
-			errorChan <- fmt.Errorf("got an EOF while reading JMX tool output")
+
+		b, err = scanner.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			queryErrC <- fmt.Errorf("reading nrjmx stdout: %s", err.Error())
+		}
+		out <- b
+		if err == io.EOF {
+			return
 		}
 	}
 }
 
 // Query executes JMX query against nrjmx tool waiting up to timeout (in milliseconds)
-func Query(objectPattern string, timeout int) (map[string]interface{}, error) {
-	defer flushWarnings()
+func Query(objectPattern string, timeoutMillis int) (result map[string]interface{}, err error) {
+	defer Close()
```

Comment:
This appears to close the process after every query -- does this mean that we need to spin up a new java process for every JMX query now?

Reply:
It was like that before, `Close` was called after receiving first result, see https://github.com/newrelic/nri-jmx/pull/47/files#diff-bbc8660333f1998a8cf44b84b3f4d8ddL262

Reply:
Oh, interesting. I was under the impression that the process stayed open until Close() was called by the integration or an error occured. Does that mean Open() needs to be called before every Query()?

Reply:
I guess it should be like that also before this as Close is always been triggered within Query. 

We didn't want to add breaking changes by modifying behaviour. Whole `jmx` package is poorly designed and I guess Close call should be up to the user to manage and not automatically triggered but that will generate a BC.

Reply:
Makes sense. That might explain some errors we were occasionally seeing. I'll have to review how we're using the package in the Kafka integration. Thanks for the explanation

Reply:
I think we could create a first SDK release without adding this BC, and then remove Close call and release again reflecting this new behaviour.

Reply:
We realised old behaviour auto-closed on empty line and timeout only, so I restored it at https://github.com/newrelic/infra-integrations-sdk/pull/200

Reply:
Restored at https://github.com/newrelic/nri-jmx/pull/47/commits/c698d75407d10c44202daaa454929d944baf21c6